Update Diagnostics API guide for native applications (#1438)
Samsung/tizen-docs: docs/application/native/guides/device/diagnostics.md

   #include <diagnostics.h>
   ```

## Diagnostic events

> Diagnotics API allow you to send and receive diagnostic events. The events can contain additional data packed into bundle.

Reviewer: ```suggestion
Diagnostics API allows you to send and receive diagnostic events. The events can contain additional data packed into the bundle.
```

Questions is it Diagnostics or Diagnostic?

To send diagnostic event, use `diagnostics_send_event()`:

```cpp
int main(int argc, char *argv[])
{
    bundle *event_data = bundle_create();

    // This is required for system services only, not applications
    diagnostics_set_client_id("diagnostics-client");

    bundle_add_str(event_data, "somekey", "somevalue");
    diagnostics_send_event("ConnectionBroken", event_data);

    bundle_free(event_data);
}
```

> To receive diagnostic evnet, use `diagnostics_subscribe_event()` proceded by `diagnostics_set_notification_cb()`:

Reviewer: event and proceeded spelling kindly fix it.

```suggestion
To receive diagnostic event, use `diagnostics_subscribe_event()` proceeded by `diagnostics_set_notification_cb()`:
```

```cpp
void notification_handler(diagnostics_ctx_h ctx, void *user_data)
{
    ...

    diagnostics_destroy(ctx);
}

int main(int argc, char **argv)
{
    diagnostics_set_notification_cb(notification_handler, NULL);
    diagnostics_subscribe_event("ConnectionBroken", "org.tizen.someapp");
}
```

## Diagnostic data request

> Diagnostics API also can request data from other application or service, especially the diagnostic data for various reasons cannot or must not be saved in the logs, such as binary or in custom format data, large volume data, reports that are time-consuming to create, and so on.

Reviewer: This is a bit confusing:
especially the diagnostic data for various reasons cannot or must not be saved in the logs, such as binary or in custom format data, large volume data, reports that are time-consuming to create, and so on.

such as binary or in custom format data it cannot be saved? or it can be requested ?

Author: The point is that with diagnostics API it is possible to request for data that has not been logged anywhere so far because its format did not allow it (binary data, large amounts of data, etc.).
Now any data (including but not limited to binary) can be retrieved using diagnostics API

Reviewer: @mmoscicki2, would this be better:

Diagnostics API also can request data from other applications or services, especially you can request the diagnostic data that could not be saved in the logs due to various reasons. For example, if the volume of data is too large, the data format is not supported as it is in binary or custom format, reports that are time-consuming to create, and so on, such data can now be requested using this API.

}
```

If `diagnostics_request_client_data()` returns successfully, the obtained data can be read by `diagnostics_data_read()`.

> Sometimes there is a need for getting data related to some specific diagnostics event. Ex. bugreport-service sends an event that a new bug report has been created and client wants to get the bug report's content after receiving the event. `diagnostics_get_data()` should be used in such case if application supports event-specific data request.

Reviewer: ```suggestion
Sometimes there is a need for getting data related to some specific diagnostics event. For example, Bugreport service sends an event that a new bug report has been created and the client wants to get the bug report's content after receiving the event. `If the application supports event-specific data request, then diagnostics_get_data()` must be used.
```

# Current implementations

## Bugreport service

> Bugreport service provides diagnostic interface to allow you to read the bug reports in case the process in the system fails. To do so, you must register the callback that will be called when the bug report is created:

Reviewer: ```suggestion
Bugreport service provides a diagnostic interface to allow you to read the bug reports in case the process in the system fails. To do so, you must register the callback `diagnostics_set_notification_cb`, it is called when the bug report is created:
```

In case a new bug report is created, callback is called. To read the content of the bug report use `diagnostics_get_data()`:

```cpp
void notification_handler(diagnostics_ctx_h ctx, void *user_data)
{
    diagnostics_data_h data;
    const char *params[] = {"--type", "crash-info-json"};

    diagnostics_get_data(ctx, params, 2, &data);

    ...

    diagnostics_destroy(ctx);
}
```

Use the `param` argument to specify which data to read from the bug report. Currently, the proper values are:

-   `--type <report_type>` - Specify the report type. Available types are:
    -   `"bugreport"` - Bugreport file is returned, which contains information about system such as system logs, list of processes, and so on. The output can be modified by dump_systemstate configuration (default).
    -   `"crash-info"` - Returned report contains information about crashed process such as callstack, mapped memory regions, and so on.
    -   `"crash-info-json"` - crash-info report in JSON format.
    -   `"fullreport"` - Full crash report ZIP archive is collected (if single report specified).
    -   `"coredumptar"` - Coredump file in TAR archive (if single report specified).

> -   `--last <seconds>` - Get reports generated in the last <seconds>

Reviewer: Does it mean to get reports generated a few seconds ago and the count of seconds is provided?

Get reports generated with the last X seconds.

}
```

If `diagnostics_request_client_data()` returns successfully, the obtained data can be read by `diagnostics_data_read()`.

Sometimes there is a need for getting data related to some specific diagnostics event. For example, Bugreport service sends an event that a new bug report has been created and the client wants to get the bug report's content after receiving the event. If the application supports event-specific data request, then `diagnostics_get_data()` must be used.

> # Current implementations

Reviewer: @mijin85cho, thanks for pointing it out. 
@mmoscicki2, could you fix this, I think there is no need for the heading Current Implementations itself however, if you feel it is needed then please modify it as @mijin85cho has suggested.

In case a new bug report is created, callback is called. To read the content of the bug report use `diagnostics_get_data()`:

```cpp
void notification_handler(diagnostics_ctx_h ctx, void *user_data)
{
    diagnostics_data_h data;
    const char *params[] = {"--type", "crash-info-json"};

    diagnostics_get_data(ctx, params, 2, &data);

    ...

    diagnostics_destroy(ctx);
}
```

Use the `param` argument to specify which data to read from the bug report. Currently, the proper values are:

-   `--type <report_type>` - Specify the report type. Available types are:
    -   `"bugreport"` - Bugreport file is returned, which contains information about system such as system logs, list of processes, and so on. The output can be modified by dump_systemstate configuration (default).
    -   `"crash-info"` - Returned report contains information about crashed process such as callstack, mapped memory regions, and so on.
    -   `"crash-info-json"` - crash-info report in JSON format.
    -   `"fullreport"` - Full crash report ZIP archive is collected (if single report specified).
    -   `"coredumptar"` - Coredump file in TAR archive (if single report specified).

> -   `--last <seconds>` - Get reports generated with the last <seconds>

Reviewer: sorry I missed within

`Get reports generated within the last <seconds>.`

Reviewer: @mijin85cho, as you can see in my comment at that time there was no period or seconds missing, it is been updated post my comment I suppose.

> -   `--from <timestamp>` - Get reports generated after specified time

Reviewer: ```suggestion
-   `--from <timestamp>` - Get reports generated after specified time.
```

> -   `--to <timestamp>` - Get reports generated before specified time (default: current time)

Reviewer: ```suggestion
-   `--to <timestamp>` - Get reports generated before specified time (default: current time).
```

-   `<report_ident>` - Get one specific report

The obtained data is associated with the `data` Diagnostics handler. To read the data content, use `diagnostics_data_read()`:

```cpp
void notification_handler(diagnostics_ctx_h ctx, void *user_data)
{
    diagnostics_data_h data;
    const char *params[] = {"--type", "fullreport"};

    int ret = diagnostics_get_data(ctx, params, 2, &data);
    if (ret != DIAGNOSTICS_ERROR_NONE) {
        diagnostics_destroy(ctx);
        return;
    }

    char buff[0x1000];
    int timeout_ms = 0;
    size_t bytes_read;

    for (;;) {
        ret = diagnostics_data_read(data, buff, sizeof(buff), timeout_ms, &bytes_read);
        if (ret ==  DIAGNOSTICS_ERROR_TRY_AGAIN)
            continue;

        if (ret != DIAGNOSTICS_ERROR_NONE)
            break; // error occurred

        if (bytes_read == 0)
            break; // no more data to read

        fwrite(buff, 1, bytes_read, stdout); // write received data to the STDOUT
    }

    diagnostics_destroy(ctx);
}
```

## Stability monitor

> Stability monitor provides interface `org.tizen.stability-monitor` for getting process specific data such as:

Reviewer: ```suggestion
Stability monitor provides interface `org.tizen.stability-monitor` for getting process-specific data such as:
```

        fwrite(buff, 1, bytes_read, stdout); // write received data to the STDOUT
    }

    diagnostics_destroy(ctx);
}
```

## Stability-monitor

> Stability monitor provides interface `org.tizen.stability-monitor` for getting process specific data such as:

Reviewer: Here it is Stability monitor,
in line 218 it is with a hyphen ## Stability-monitor

Can we avoid the hyphen, I understand there is a hyphen here `org.tizen.stability-monitor` but consistency sake either use with hyphen in heading and sentence or without.
